MRG: Ica inf values fix

ARudiuk opened the pull request:
It seems like the error with #3267 is that the user can define channels which are not supported data channels.

In my current code I have added a simple method that crashes the program if there are unsupported channels defined by the user. I also added EOG as an acceptable channel. 
@agramfort @Eric89GXL any ideas on how to proceed? 

Two things that still need to be addressed in my eyes are: 1) should EOG be included as a supported channel? 2) how would I raise an appropriate error for bad channels being present?

ARudiuk commented:
Here is code if you want to crash the program without my fix,
```
import numpy as np

import mne
from mne.datasets import sample
from mne.utils import set_log_level

from mne.preprocessing import ICA, read_ica
from mne.preprocessing import create_ecg_epochs, create_eog_epochs

data_path = sample.data_path()
raw_fname = data_path + '/MEG/sample/sample_audvis_filt-0-40_raw.fif'
raw = mne.io.read_raw_fif(raw_fname, preload=True)
raw.filter(1, 45, n_jobs=2)

picks_eeg = mne.pick_types(raw.info, meg=False, eeg=True, eog=True,
                           stim=False, exclude='bads')
#reject = dict(mag=4e-12, grad=4000e-13)  # don't fit on crazy environmental artefacts
# ICA parameters
n_components = 5  # if float, select the number components by explained variance of PCA
method = 'fastica'  # for comparison with EEGLAB try "extended-infomax" here
decim = 3  # we only need enough statistics, not all time points
# Define the ICA object instance
ica = ICA(n_components=n_components, method=method)
# reject = dicteeg=3*10-2)  # don't fit on crazy environmental artefacts
ica.fit(raw, picks=picks_eeg, decim=decim)
```

Eric89GXL commented:
Can you make a unit test out of it (fails on master, passes on your branch)
using the "testing" data?

Eric89GXL commented:
See how other tests in mne/preprocessing/tests/ do it

agramfort commented on mne/preprocessing/ica.py:
```diff
@@ -270,6 +271,13 @@ def __repr__(self):
+        n_chan = info['nchan']
+        check = all([channel_type(info, j) in types for j in picks])
+        if not check:
+            1 / 0
```
hum.... I would expect a proper ValueError exception here

agramfort commented on mne/preprocessing/ica.py:
```diff
@@ -440,16 +450,18 @@ def _pre_whiten(self, data, info, picks):
                     elif ch_type == 'eeg':
                         this_picks = pick_types(info, meg=False, eeg=True)
-                    else:
+                    elif ch_type == 'mag' or ch_type == 'grad':
```
elif ch_type in ['mag', 'grad']:

agramfort commented:
@ARudiuk you should not merge the master branch into your feature branches. Never do:

git merge master

it makes the history of the code really hard to read.

agramfort commented:
let us know when you addressed the comments.

@jona-sassenhagen with this you should be able to run ICA keeping the EOG as eog channels

ARudiuk commented:
@agramfort Sorry, I missed the the re-base section. Should I try to fix it now, or leave it and not do it in the future.

agramfort commented:
See if you can fix it eventually rewriting one of the commits 






On Sun, May 29, 2016 at 1:13 PM -0700, "Alexander Rudiuk" <notifications@github.com> wrote:












@agramfort Sorry, I missed the the re-base section. Should I try to fix it now, or leave it and not do it in the future. 



—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

ARudiuk commented:
I addressed the comments. Primarily the unit test, and re-basing to remove the merge.

The unit test checks raw and epoched data for a bad channel ('stim') and checks if it crashes when proper channels are present/absent. 

@agramfort @Eric89GXL

agramfort commented on mne/preprocessing/ica.py:
```diff
@@ -270,6 +271,18 @@ def __repr__(self):
+        """Check for ica channels that are not considered
+        actual data channels. These prevents the program from
+        crashing without feedback when a bad channel is provided
+        to ICA whitening
```
be more specific. Mention that it allows EOG channels as well as data channels in ICA.

agramfort commented on mne/preprocessing/ica.py:
```diff
@@ -270,6 +271,18 @@ def __repr__(self):
+        check = all([channel_type(info, j) in types for j in picks])
+        if not check:
+            raise ValueError("""Invalid channel type(s) passed for ICA.
+             Only the following channels are supported {}""".format(types))
```
bad indent for pep8

agramfort commented on mne/preprocessing/tests/test_ica.py:
```diff
@@ -609,4 +609,32 @@ def test_fit_params():
+
+def test_bad_channels():
+    """Test if exception is raised when unsupported channels are used"""
+    """Test for raw and evoked data"""
```
"""" is for multiline strings. No need to have """ for every line.

agramfort commented on mne/preprocessing/tests/test_ica.py:
```diff
@@ -609,4 +609,32 @@ def test_fit_params():
+def test_bad_channels():
+    """Test if exception is raised when unsupported channels are used"""
+    """Test for raw and evoked data"""
+    raw = Raw(raw_fname)
```
just do `Raw(raw_fname, preload=True)` and remove line below.

agramfort commented on mne/preprocessing/tests/test_ica.py:
```diff
@@ -609,4 +609,32 @@ def test_fit_params():
+    picks_bad2 = pick_types(epochs.info, meg=True, stim=True, ecg=True,
+                            eog=True, exclude='bads')
+    assert_raises(ValueError, ica.fit, epochs, picks=picks_bad1, decim=decim)
+    assert_raises(ValueError, ica.fit, epochs, picks=picks_bad2, decim=decim)
```
I think you can write this with a for loop

```
for inst in [raw, epochs):
   ... = pick_types(inst.info, ...)
   ... = ...
```

Eric89GXL commented on mne/preprocessing/ica.py:
```diff
@@ -440,16 +455,18 @@ def _pre_whiten(self, data, info, picks):
+                    elif ch_type in ['mag', 'grad']:
                         this_picks = pick_types(info, meg=ch_type)
+                    elif ch_type == 'eog':
+                        this_picks = pick_types(info, meg=False, eog=True)
```
just for safety you should put something like:
```
else:
    raise RuntimeError('Unclassified channel type %s' % ch_type)
```

jaeilepp replied:
Should these functions have ``ref_meg=False`` too?

ARudiuk replied:
Sorry, I'm not familiar with ref_meg. I'll have to look into what it is before I know

agramfort replied:
just test some CTF data eg. spm dataset to see if it works with ref_meg
channels

ARudiuk replied:
The code catches ref_meg and says it is not a supported channel. Is that the desired behaviour? @Eric89GXL @agramfort @jaeilepp

ARudiuk replied:
Should I write a test case that checks if it is catched?

agramfort replied:
yes It's expected. A test with a manual setting of a channel as ref_meg
would be nice indeed.

ARudiuk commented:
@agramfort @Eric89GXL committed changes

Eric89GXL commented on mne/preprocessing/ica.py:
```diff
@@ -270,6 +271,22 @@ def __repr__(self):
+        """
+        types = _DATA_CH_TYPES_SPLIT + ['eog']
+        check = all([channel_type(info, j) in types for j in picks])
+        if picks == []:
```
if `len(picks) == 0` is probably better

Eric89GXL commented on mne/preprocessing/ica.py:
```diff
@@ -270,6 +271,22 @@ def __repr__(self):
+        feedback when a bad channel is provided to ICA whitening.
+        """
+        types = _DATA_CH_TYPES_SPLIT + ['eog']
+        check = all([channel_type(info, j) in types for j in picks])
```
you could instead do:
```
unknown = sorted(list(set(channel_type(info, pick) for pick in picks) - set(types)))
```
And then do
```
if len(unknown) > 0:
    raise ValueError('..., got %s' % unknown)
```

ARudiuk replied:
Why is the list being sorted? Is it better to do this through sets than checking every list element?

Eric89GXL replied:
The set should take around the same time but make for a nicer error message. The sorting is also done for the `ValueError` message. You could put it there, but if the list is empty, the `sorted` will return almost immediately so it should be fine. And actually `sorted` constructs a list so you can even just do:
```
unknown = sorted(set(channel_type(info, pick) for pick in picks) - set(types))
```

Eric89GXL commented on mne/preprocessing/ica.py:
```diff
@@ -270,6 +271,22 @@ def __repr__(self):
 
         return '<ICA  |  %s>' % s
 
+    def _check_for_unsupported_ica_channels(self, picks, info):
```
this doesn't use `self` at all, so it should be a small helper function instead of a method

ARudiuk replied:
Does this just mean moving it outside the class?

agramfort replied:
Yes






On Thu, Jun 2, 2016 at 9:10 PM -0700, "Alexander Rudiuk" <notifications@github.com> wrote:












In mne/preprocessing/ica.py:
> @@ -270,6 +271,22 @@ def __repr__(self):
>  
>          return '<ICA  |  %s>' % s
>  
> +    def _check_for_unsupported_ica_channels(self, picks, info):



Does this just mean moving it outside the class?



—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

Eric89GXL commented on mne/preprocessing/ica.py:
```diff
@@ -326,6 +343,8 @@ def fit(self, inst, picks=None, start=None, stop=None, decim=None,
         self : instance of ICA
             Returns the modified instance.
         """
+        if picks is not None:
```
this conditional can go inside the helper checking function

Eric89GXL commented on mne/preprocessing/tests/test_ica.py:
```diff
@@ -608,4 +608,27 @@ def test_fit_params():
 
+
+@requires_sklearn
+def test_bad_channels():
```
You added functionality that allows EOG to be excluded, right? If so please add a test for it to make sure it at least runs without error

ARudiuk replied:
Should it be a test that checks if every good channel works?

ARudiuk replied:
Then should I test if every bad channel crashes?

Eric89GXL replied:
Sorry, I meant EOG to be *included*. You should add a test that makes sure you can include EOG channels in the ICA process, since that's what this PR adds (right?). At the very least just make sure it works without throwing an error. This is the type of test that would throw an error on `master` currently, so by including it you're working on TDD (test-driven development) -- although strictly speaking I think such a test should be the *first* change, not the last, to qualify as TDD. But in practice we rarely do it that way :)

ARudiuk replied:
I added a new test method that simply calls `ica.fit` when an EOG channel is present. The method feels a bit off compared to the other test ones that are implemented, because it does not call an `assert` anywhere. The reason for that is that I don't know what I would compare the resulting value with since ICA is not deterministic right?

Eric89GXL replied:
You can fix the random_state, see how it's done in other ICA tests. I'm not
sure what exactly you'd want to test, maybe that it's actually including
the EOG channel or something. Maybe the other existing ICA tests will give
you some ideas.

agramfort replied:
Check eog is in ica.ch_names






On Sat, Jun 4, 2016 at 1:02 PM -0700, "Eric Larson" <notifications@github.com> wrote:












In mne/preprocessing/tests/test_ica.py:
> @@ -608,4 +608,27 @@ def test_fit_params():
>      ICA(fit_params=fit_params)  # test no side effects
>      assert_equal(fit_params, {})
>  
> +
> +@requires_sklearn
> +def test_bad_channels():

You can fix the random_state, see how it's done in other ICA tests. I'm not
sure what exactly you'd want to test, maybe that it's actually including
the EOG channel or something. Maybe the other existing ICA tests will give
you some ideas.



—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

agramfort commented:
@ARudiuk when you addressed all comments set your PR to MRG (replace WIP by MRG in title)

agramfort commented:
@ARudiuk can you make travis happy?

ARudiuk commented:
@agramfort Will do, just getting ready to head home from vacation today so need to find some time first.

ARudiuk commented:
Is there a fake dataset that has every possible channel type so I can loop through all the bad ones as a test?

agramfort commented:
no you need to make one artifically by using set_channel_types method on
our test raw

agramfort commented:
ready on your side?

ARudiuk commented:
Yup

agramfort commented on mne/preprocessing/tests/test_ica.py:
```diff
@@ -608,4 +610,58 @@ def test_fit_params():
+        for ch in chs_bad:
+            # Test case for only bad channels
+            picks_bad1 = pick_types(inst.info, exclude='bads',
+                                    **{str(ch): True})
```
sweet pythonic trick :)

agramfort commented on mne/preprocessing/tests/test_ica.py:
```diff
@@ -608,4 +610,58 @@ def test_fit_params():
+        picks1 = pick_types(inst.info, meg=True, stim=False, ecg=False,
+                            eog=False, exclude='bads')
+        ica.fit(inst, picks=picks1)
+        assert_false(any('EOG' in ch for ch in ica.ch_names))
```
how long does it take to run the 2 added tests? Can you make it even faster? cropping the raw, using fastica?

ARudiuk replied:
It took 8 seconds, but I just got it down to 0.336 seconds.

jaeilepp commented on mne/preprocessing/tests/test_ica.py:
```diff
@@ -608,4 +610,58 @@ def test_fit_params():
+                                    **{str(ch): True})
+            # Test case for good and bad channels
+            picks_bad2 = pick_types(inst.info, exclude='bads',
+                                    **{str(ch): True})
```
Why are you doing this twice? Should this have ``exclude=[]``?

ARudiuk replied:
Oops, one of them should include an MEG channel.

ARudiuk replied:
There are no bad channels in that data set, so I can probably just drop that

Eric89GXL replied:
Or you can always add some "bad channels" (i.e., in reality fine but useful to mark as bad for testing) to `inst.info['bads']` if it's worth tetsing the functionality

ARudiuk replied:
I'll keep that in mind

ARudiuk commented:
@agramfort fixed up test cases

ARudiuk commented:
Closes #3267

agramfort commented on mne/preprocessing/ica.py:
```diff
@@ -440,16 +464,22 @@ def _pre_whiten(self, data, info, picks):
+                        this_picks = pick_types(info, meg=False, eog=True)
+                    else:
+                        raise RuntimeError("""Should not be reached.
+                                           Unsupported channel {}"""
```
we use """ just for docstrings

agramfort commented:
besides +1 for merge

ARudiuk commented:
Should I add anything to the what's new document? @agramfort @Eric89GXL

agramfort commented:
yes please in API.

ARudiuk commented:
Added

agramfort commented:
thanks @ARudiuk
